types/reply.d.ts

```diff
@@ -1,42 +1,52 @@
   RawReply extends RawReplyDefaultExpression<RawServer> = RawReplyDefaultExpression<RawServer>,
-  ContextConfig = ContextConfigDefault
+  RequestGeneric extends RequestGenericInterface = RequestGenericInterface,
+  ContextConfig = ContextConfigDefault,
 > {
-  callNotFound(): void;
-  code(statusCode: number): FastifyReply<RawServer, RawReply>;
-  hasHeader(key: string): boolean;
-  header(key: string, value: any): FastifyReply<RawServer, RawReply>;
+  raw: RawReply;
+  context: FastifyContext<ContextConfig>;
+  log: FastifyLoggerInstance;
+  request: FastifyRequest<RawServer, RawRequest, RequestGeneric>;
+  code(statusCode: number): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  status(statusCode: number): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  statusCode: number;
+  sent: boolean;
+  send<T>(payload?: T): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  header(key: string, value: any): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  headers(values: {[key: string]: any}): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
   getHeader(key: string): string | undefined;
   getHeaders(): {
     // Node's `getHeaders()` can return numbers and arrays, so they're included here as possible types.
     [key: string]: number | string | string[] | undefined;
   };
-  // Note: should consider refactoring the argument order for redirect. statusCode is optional so it should be after the required url param
-  redirect(statusCode: number, url: string): FastifyReply<RawServer, RawReply>;
-  redirect(url: string): FastifyReply<RawServer, RawReply>;
   removeHeader(key: string): void;
-  send<T>(payload?: T): FastifyReply<RawServer, RawReply>;
-  serialize(payload: any): string;
-  serializer(fn: (payload: any) => string): FastifyReply<RawServer, RawReply>;
-  status(statusCode: number): FastifyReply<RawServer, RawReply>;
-  statusCode: number;
+  hasHeader(key: string): boolean;
+  // Note: should consider refactoring the argument order for redirect. statusCode is optional so it should be after the required url param
+  redirect(statusCode: number, url: string): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  redirect(url: string): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  callNotFound(): void;
   getResponseTime(): number;
-  type(contentType: string): FastifyReply<RawServer, RawReply>;
-  context: FastifyContext<ContextConfig>;
-  raw: RawReply;
-  sent: boolean;
+  type(contentType: string): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  serializer(fn: (payload: any) => string): FastifyReply<RawServer, RawRequest, RawReply, RequestGeneric, ContextConfig>;
+  serialize(payload: any): string;
+  then(fullfilled: () => void, rejected: (err: Error) => void): void;
```

Comment:
shouldn't this return another `thenable`?

Reply:
You're right about the spec, but I think it is what the code does right now.
https://github.com/fastify/fastify/blob/master/lib/reply.js#L299

Reply:
should open a new issue about this if this is a concern